### ISSUE TITLE
BUG: non-geometry column named "geometry" gets discarded when using gpd.GeoDataFrame constructor

### ISSUE THREAD

**Issue by phildias** [geopandas/geopandas#2312]
- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of geopandas.
- [ ] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd
import geopandas as gpd

df = pd.DataFrame({'x':[0,1],'y':[0,1],'id':[1,2],'geometry':['a','b']})
print(df)
#    x  y  id geometry
# 0  0  0   1        a
# 1  1  1   2        b

gdf = gpd.GeoDataFrame(df, geometry=gpd.points_from_xy(df['x'], df['y']), crs='epsg:4326')
print(gdf)
#    x  y  id                 geometry
# 0  0  0   1  POINT (0.00000 0.00000)
# 1  1  1   2  POINT (1.00000 1.00000)
```

#### Problem description

In `df`, the input pd.DataFrame, the column called "geometry" with some string values. However, when it was passed to the `gpd.GeoDataFrame()` constructor method, that column got over-written by the column containing the geometry.

#### Expected Output

The `gpd.GeoDataFrame()` constructor method should look at the already-existing columns of the input pd.DataFrame and, if the name "geometry" is already taken, just keep tacking on underscores onto "`geometry`" until it finds an unused column name.

#### Output of ``geopandas.show_versions()``

<details>

SYSTEM INFO
-----------
python     : 3.9.9 | packaged by conda-forge | (main, Dec 20 2021, 02:36:06) [MSC v.1929 64 bit (AMD64)]
executable : C:\ProgramData\Anaconda3\envs\geo_env_\python.exe
machine    : Windows-10-10.0.19042-SP0

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : None
GEOS lib   : None
GDAL       : 3.4.0
GDAL data dir: None
PROJ       : 8.2.0
PROJ data dir: C:\ProgramData\Anaconda3\envs\geo_env_\Library\share\proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.10.2
pandas     : 1.3.5
fiona      : 1.8.20
numpy      : 1.22.0
shapely    : 1.8.0
rtree      : 0.9.7
pyproj     : 3.3.0
matplotlib : 3.5.1
mapclassify: 2.4.3
geopy      : 2.2.0
psycopg2   : 2.9.3 (dt dec pq3 ext lo64)
geoalchemy2: None
pyarrow    : None
pygeos     : 0.12.0

</details>

### Extra notes

I believe this issue is very similar to #1964 and #2310 .

**Comment by m-richards** [comment]
Thanks for reporting this as well.

Strictly speaking, this is actually documented behaviour;
https://github.com/geopandas/geopandas/blob/831108dcb369517e311cd0ee3821db1f2fb7122d/geopandas/geodataframe.py#L90-L92
but the result is less than ideal in the case you've demonstrated.

I think however in the case this occurs, raising a warning or exception in this case would probably be more appropriate than aliasing the geometry column with underscores.

This isn't likely to occur very often and when it does, it seems plausible many of those cases are accidental. Given that, I think it makes sense for a user to see a warning or error and resolve the issue, rather than implicitly "fixing" things.

**Comment by phildias** [comment]
I've been thinking about this for the last few days and I think I was able to identify why I want to change the genera behavior of the constructor. The way it currently works, the `geometry` attribute is somewhat ambiguous.

When the `data` parameter is a pd.DataFrame (or a dict or list containing DF-like data), and the `geometry` parameter is just a string, the string used in the `geometry` attribute gets used as the column name that is searched inside the `data` parameter to use as the geometry column. 

When the `data` parameter is a pd.DataFrame (or a dict or list containing DF-like data), and the `geometry` parameter is a list or gpd.GeoSeries containing shapely geometries, a new column gets added to the `data` parameter and it **ALWAYS** uses the column name "`geometry`". When the user passes a list or gpd.GeoSeries to the `geometry` parameter, they don't have the option of also determining what name should be used for that new column.

I'm not exactly sure what the best approach would be to implement a solution here. I thought about creating a few new input parameters for the constructor method such as `geometries` and `geometry_column_name` (instead of just `geometry`). But that breaks the current API and it can get super messy.

**Comment by jorisvandenbossche** [comment]
When passing a Series as geometry values, it is true that we currently ignore the `name` of the Series:

```python
>>> geometries = geopandas.GeoSeries.from_xy([0, 1], [0, 1], name="my_geom")
>>> geometries
0    POINT (0.00000 0.00000)
1    POINT (1.00000 1.00000)
Name: my_geom, dtype: geometry

>>> geopandas.GeoDataFrame({"col": [0, 1]}, geometry=geometries)
   col                 geometry
0    0  POINT (0.00000 0.00000)
1    1  POINT (1.00000 1.00000)
```

I don't know if that's something we considered before, but we should maybe preserve the Series name in this case (if it has one).

> I'm not exactly sure what the best approach would be to implement a solution here. I thought about creating a few new input parameters for the constructor method

It's correct that, when passing the geometries as values, you are less flexible in naming and we use the default "geometry" name. But, I am personally not sure that is necessarily the problem. Because if you want, the other option of specifying the column name is always available as an alternative if you first add the column to the existing dataframe with the name you prefer. 
For example, you could do something like this:

```python
gpd.GeoDataFrame(df.assign(my_geoms=gpd.points_from_xy(df['x'], df['y']), crs='epsg:4326'), geometry="my_geoms")
```

**Comment by m-richards** [comment]
>But, I am personally not sure that is necessarily the problem

I would agree with that as well - `gpd.GeoDataFrame(...).rename_geometry("my_geoms")` is another way of doing this. While in principle we could add additional keyword args to do this inside the init, I think it potentially makes it more confusing to use - especially for someone new to the library.

**Comment by phildias** [comment]
Ooohhh, I really like the idea of making use of the GeoSeries' `name` attribute when it exists. 

If we're going to make use of the `name` attribute of the GeoSeries, I think it would also be beneficial for the `points_from_xy` method to accept a `name` parameter/input. 

Both of these modifications (using the GeoSeries' `name` attribute when making a new GeoDataFrame and making the `points_from_xy` method accept a new `name` parameter) would likely require editing the following pieces of code:
- The `points_from_xy` method inside `geopandas/array.py`: https://github.com/geopandas/geopandas/blob/831108dcb369517e311cd0ee3821db1f2fb7122d/geopandas/array.py#L215
- The `GeometryArray` class inside `geopandas/array.py`: https://github.com/geopandas/geopandas/blob/831108dcb369517e311cd0ee3821db1f2fb7122d/geopandas/array.py#L261
- The `GeoDataFrame.__init__` method inside `geopandas/geodataframe.py`: https://github.com/geopandas/geopandas/blob/13bec536192a48866334f3b3833e84817cb0fb65/geopandas/geodataframe.py#L134

I'll start messing around with these and report back when I have something functional. 

Thanks for the feedback with all this!

PS: On an unrelated note, you mentioned: 
> For example, you could do something like this:
> ```python
> gdf = gpd.GeoDataFrame(
>     df.assign(my_geoms=gpd.points_from_xy(df['x'], df['y']), crs='epsg:4326'), 
>     geometry="my_geoms"
> )
> ```
The code above adds a column called "crs" with all rows containing the value `epsg:4326`.  I think you meant something like this:
```python
gdf = gpd.GeoDataFrame(
    df.assign(my_geoms=gpd.points_from_xy(df['x'], df['y'])),
    crs='epsg:4326',
    geometry="my_geoms")
```

**Comment by martinfleis** [comment]
This is also related to #1038.
